### DISCUSSION TITLE
What is the right abstraction for the line number mark?

### DISCUSSION THREAD

**Discussion by pokey** [cursorless-dev/cursorless#1133] [Q&A]
# Background

We have three different types of line number:

- Absolute line numbers modulo 100.  Eg `"take row twenty"` will take the only visible line that is 20 modulo 100
- Absolute line numbers.  There is actually no spoken form for these because it would explode grammar
- Relative line numbers, eg `"take up five"`.  These are relative to the "primary cursor" when you have multiple cursors.  This corresponds to the way VSCode shows the line numbers when you have the `"editor.lineNumbers": "relative"` setting
  
<img width="272" alt="image" src="https://user-images.githubusercontent.com/755842/201935302-5ead4b7f-6840-4195-9767-4f7d5351cb1d.png">

# The question

What is the right abstraction that will be portable across editors?  Do other editors have relative line numbers?  Do they implement them the same way (ie relative to the "primary" cursor)?  Do other editors even have a notion of the "primary" cursor?

# Relevant links

- https://github.com/cursorless-dev/cursorless/pull/1128#discussion_r1021906747

cc/ @rntz @phillco @JohnEfford @dyamito

**Comment by JohnEfford** [comment]
Visual studio displays the absolute line number, there is a relative line number extension but it does not seem to available for the current version, I just checked settings and there is no relative mode available natively. I’m comfortable using the modulo line numbers for selections, without relative line numbers I only use relative selections for small selections scopes. I would say stick with Modulo 100 for the spoken form of row, the only conceivable problem being people with portrait mode monitors that display more than 100 lines, in which case take row, should no longer work.

Cheers
John
